phoenix-core/src/main/java/org/apache/phoenix/schema/transform/TransformMaintainer.java

```diff
@@ -0,0 +1,459 @@
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e); // Impossible
+            }
+        }
+        return maintainers;
+    }
+
+    // Return new table's name
+    public byte[] getIndexTableName() {
+        return newTableName;
+    }
+
+    // Builds new table's rowkey using the old table's rowkey.
+    // This method will change when we support rowkey related transforms
+    public byte[] buildRowKey(ValueGetter valueGetter, ImmutableBytesWritable rowKeyPtr, byte[] regionStartKey, byte[] regionEndKey, long ts)  {
+        ImmutableBytesWritable ptr = new ImmutableBytesWritable();
+        boolean isNewTableSalted = nNewTableSaltBuckets > 0;
+        TrustedByteArrayOutputStream stream = new TrustedByteArrayOutputStream(estimatedNewTableRowKeyBytes);
+        DataOutput output = new DataOutputStream(stream);
+
+        try {
+            if (isNewTableSalted) {
+                output.write(0); // will be set at end to new table salt byte
+            }
+            // The oldTableRowKeySchema includes the salt byte field,
+            // so we must adjust for that here.
+            int dataPosOffset = isOldTableSalted ? 1 : 0 ;
+            //BitSet viewConstantColumnBitSet = this.rowKeyMetaData.getViewConstantColumnBitSet();
+            // Skip data table salt byte
+            int maxRowKeyOffset = rowKeyPtr.getOffset() + rowKeyPtr.getLength();
+            oldTableRowKeySchema.iterator(rowKeyPtr, ptr, dataPosOffset);
+
+            // Write new table row key
+            while (oldTableRowKeySchema.next(ptr, dataPosOffset, maxRowKeyOffset) != null) {
+                output.write(ptr.get(), ptr.getOffset(), ptr.getLength());
+                if (!oldTableRowKeySchema.getField(dataPosOffset).getDataType().isFixedWidth()) {
+                    output.writeByte(SchemaUtil.getSeparatorByte(newTableRowKeyOrderOptimizable, ptr.getLength()==0
+                            , oldTableRowKeySchema.getField(dataPosOffset)));
+                }
+                dataPosOffset++;
+            }
+
+            byte[] newTableRowKey = stream.getBuffer();
+            // Remove trailing nulls
+            int length = stream.size();
+            if (isNewTableSalted) {
+                // Set salt byte
+                byte saltByte = SaltingUtil.getSaltingByte(newTableRowKey, SaltingUtil.NUM_SALTING_BYTES, length-SaltingUtil.NUM_SALTING_BYTES, nNewTableSaltBuckets);
+                newTableRowKey[0] = saltByte;
+            }
+            return newTableRowKey.length == length ? newTableRowKey : Arrays.copyOf(newTableRowKey, length);
+        } catch (IOException e) {
+            throw new RuntimeException(e); // Impossible
+        } finally {
+            try {
+                stream.close();
+            } catch (IOException e) {
+                throw new RuntimeException(e); // Impossible
```

Comment:
Good to wrap `stream` in try-with-resource to avoid having to handle such possibilities here.